x-frame-options/support/helper.sub.js

```diff
@@ -0,0 +1,65 @@
+function xfo_simple_tests({ headerValue, headerValue2, cspValue, sameOriginAllowed, crossOriginAllowed }) {
+  const value2QueryString = headerValue2 !== undefined ? `&value2=${headerValue2}` : ``;
+  const cspQueryString = cspValue !== undefined ? `&csp_value=${cspValue}` : ``;
+
+  const valueMessageString = headerValue === "" ? "(the empty string)" : headerValue;
+  const value2MessageString = headerValue2 !== undefined ? `;${headerValue2}` : ``;
+  const cspMessageString = cspValue !== undefined ? ` with CSP ${cspValue}` : ``;
+
+  xfo_test({
+    url: `/x-frame-options/support/xfo.py?value=${headerValue}${value2QueryString}${cspQueryString}`,
+    check: sameOriginAllowed ? "loaded message" : "no message",
+    message: `${valueMessageString}${value2MessageString} ${sameOriginAllowed ? "allows" : "blocks"} same-origin framing${cspMessageString}`
+  });
+
+  xfo_test({
+    url: `http://{{domains[www]}}:{{ports[http][0]}}/x-frame-options/support/xfo.py?value=${headerValue}${value2QueryString}${cspQueryString}`,
+    check: crossOriginAllowed ? "loaded message" : "no message",
+    message: `${valueMessageString}${value2MessageString} ${crossOriginAllowed ? "allows" : "blocks"} cross-origin framing${cspMessageString}`
+  });
+}
+
+function xfo_test({ url, check, message }) {
+  async_test(t => {
+    const i = document.createElement("iframe");
+    i.src = url;
+
+    switch (check) {
+      case "loaded message": {
+        waitForMessageFrom(i, t).then(t.step_func_done(e => {
+          assert_equals(e.data, "Loaded");
+        }));
+        break;
+      }
+      case "failed message": {
+        waitForMessageFrom(i, t).then(t.step_func_done(e => {
+          assert_equals(e.data, "Failed");
+        }));
+        break;
+      }
+      case "no message": {
+        waitForMessageFrom(i, t).then(t.unreached_func("Frame should not have sent a message."));
+        i.onload = t.step_func_done(() => {
```

Comment:
I suspect a problem here is that Firefox does not dispatch the load event for (all) network errors. See also https://github.com/whatwg/html/issues/125 (and references to that in other tests). If initially `i.contentDocument` was available as I suspect it would be given initial about:blank, you could wait for it to not be available using [step_wait and friends](https://web-platform-tests.org/writing-tests/testharness-api.html#timeouts-in-tests).